sklearn/_fast_gradient_boosting/grower.py

```diff
@@ -0,0 +1,505 @@
+the gradients and hessians of the training data.
+"""
+from heapq import heappush, heappop
+import numpy as np
+from time import time
+
+from .splitting import Splitter
+from .predictor import TreePredictor, PREDICTOR_RECORD_DTYPE
+
+from .types import HISTOGRAM_DTYPE
+
+
+class TreeNode:
+    """Tree Node class used in TreeGrower.
+
+    This isn't used for prediction purposes, only for training (see
+    TreePredictor).
+
+    Parameters
+    ----------
+    depth : int
+        The depth of the node, i.e. its distance from the root
+    sample_indices : array of int
+        The indices of the samples at the node
+    sum_gradients : float
+        The sum of the gradients of the samples at the node
+    sum_hessians : float
+        The sum of the hessians of the samples at the node
+    parent : TreeNode or None, optional(default=None)
+        The parent of the node. None for root.
+
+    Attributes
+    ----------
+    depth : int
+        The depth of the node, i.e. its distance from the root
+    sample_indices : array of int
+        The indices of the samples at the node
+    sum_gradients : float
+        The sum of the gradients of the samples at the node
+    sum_hessians : float
+        The sum of the hessians of the samples at the node
+    parent : TreeNode or None, optional(default=None)
+        The parent of the node. None for root.
+    split_info : SplitInfo or None
+        The result of the split evaluation
+    left_child : TreeNode or None
+        The left child of the node. None for leaves.
+    right_child : TreeNode or None
+        The right child of the node. None for leaves.
+    value : float or None
+        The value of the leaf, as computed in finalize_leaf(). None for
+        non-leaf nodes
+    find_split_time : float
+        The total time spent computing the histogram and finding the best
+        split at the node.
+    apply_split_time : float
+        The total time spent actually splitting the node, e.g. splitting
+        sample_indices into left and right child.
+    hist_subtraction : bool
+        Wheter the subtraction method was used for computing the histograms.
+    start : int
+        start position of the node's sample_indices in splitter.partition
+    stop : int
+        stop position of the node's sample_indices in splitter.partition
+    """
+
+    split_info = None
+    left_child = None
+    right_child = None
+    value = None
+    histograms = None
+    sibling = None
+    parent = None
+    find_split_time = 0.
+    apply_split_time = 0.
+    hist_subtraction = False
+
+    # start and stop indices of the node in the splitter.partition
+    # array. Concretely,
+    # self.sample_indices = view(self.splitter.partition[start:stop])
+    # Only used in _update_raw_prediction, because we need to iterate over the
+    # leaves and I don't know how to efficiently store the sample_indices
+    # views because they're all of different sizes.
+    start = 0
+    stop = 0
```

Comment:
@ogrisel WDYT about this?
This is the most convenient way I found to cythonize `_update_raw_predictions` properly.

Reply:
I would rename to `partition_start` / `partition_stop` to be slightly more explicit.